### PR TITLE
OCM-2446 | feat: add a rosa describe machinepool command

### PR DESCRIPTION
**Pull request by andreadecorte**
Add a new rosa describe machinepool command that shows details for a single machine pool, with json/yaml output also available. The command is also showing any scheduled node pool upgrade for HCP clusters.

This also removes a few HCP items from rosa list machinepool command as it was becoming too cluttered. All those can now be found on the new describe command.

Related: [OCM-2446](https://issues.redhat.com//browse/OCM-2446)
Design doc: https://docs.google.com/document/d/1NlvXEHR0Q5mJEQC8Y4uIuzYLaKiaQFnLFR3iK7xDGy4

Example of outputs:
classic cluster
```
ID:                         worker
Cluster ID:                 clusterId
Autoscaling:                No
Replicas:                   2
Instance type:              m5.xlarge
Labels:
Taints:
Availability zones:         us-east-2a
Subnets:
Spot instances:             No
Disk size:                  default
```

HCP cluster with scheduled upgrades
```
ID:                         workers
Cluster ID:                 clusterId
Autoscaling:                No
Desired replicas:           2
Current replicas:           2
Instance type:              m5.xlarge
Labels:
Taints:
Availability zone:          us-east-1a
Subnet:
Version:                    4.12.24
Autorepair:                 Yes
Tuning configs:
Message:
Scheduled upgrade:          scheduled 4.12.25 on 2023-09-12 15:22 UTC
```

### CODE REVIEW

**Comment by lucasponce** [comment]
It looks nice.

A first comment would be to align the desired/current replicas in the list command as discussed in:

https://redhat-internal.slack.com/archives/G0109Q17YTH/p1682682224134909

I think this PR would be ideal.

File: cmd/describe/machinepool/machinepool.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+func describeMachinePool(r *rosa.Runtime, cluster *cmv1.Cluster, clusterKey string, machinePoolID string) error {
+	r.Reporter.Debugf("Fetching machine pool '%s' for cluster '%s'", machinePoolID, clusterKey)
+	machinePool, exists, err := r.OCMClient.GetMachinePool(cluster.ID(), machinePoolID)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		r.Reporter.Infof("Machine pool '%s' not found", machinePoolID)
```
Comment (oriAdler) [inline cmd/describe/machinepool/machinepool.go:19]: I wonder if we should error out instead.
From the describe cluster command:
```
oadler@fedora:rosa (master)$ rosa describe cluster --cluster oadler
E: Failed to get cluster 'oadler': There is no cluster with identifier or name 'oadler'
```
WDYT?

File: cmd/describe/machinepool/machinepool.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		return nil
+	}
+
+	if output.HasFlag() {
+		err = output.Print(machinePool)
```
Comment (oriAdler) [inline cmd/describe/machinepool/machinepool.go:24]: NP
```suggestion
		return output.Print(machinePool)
```

File: cmd/describe/machinepool/nodepool.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+func describeNodePool(r *rosa.Runtime, cluster *cmv1.Cluster, clusterKey string, nodePoolID string) error {
+	r.Reporter.Debugf("Fetching node pool '%s' for cluster '%s'", nodePoolID, clusterKey)
+	nodePool, exists, err := r.OCMClient.GetNodePool(cluster.ID(), nodePoolID)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		r.Reporter.Infof("Machine pool '%s' not found", nodePoolID)
```
Comment (oriAdler) [inline cmd/describe/machinepool/nodepool.go:21]: Also here, we might want to error instead.

File: cmd/describe/machinepool/nodepool.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		return err
+	}
+
+	if output.HasFlag() {
+		var formattedOutput map[string]interface{}
+		formattedOutput, err = formatNodePoolOutput(nodePool, scheduledUpgrade)
+		if err != nil {
+			return err
+		}
+		err = output.Print(formattedOutput)
```
Comment (oriAdler) [inline cmd/describe/machinepool/nodepool.go:36]: NP
```suggestion
		return output.Print(formattedOutput)
```

File: cmd/describe/machinepool/nodepool.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		nodePool.AvailabilityZone(),
+		nodePool.Subnet(),
+		ocmOutput.PrintNodePoolVersion(nodePool.Version()),
+		ocmOutput.PrintNodePoolAutorepair(nodePool.AutoRepair()),
+		ocmOutput.PrintNodePoolTuningConfigs(nodePool.TuningConfigs()),
+		ocmOutput.PrintNodePoolMessage(nodePool.Status()),
+	)
+
+	// Print scheduled upgrades if existing
+	if scheduledUpgrade != nil {
```
Comment (lucasponce) [inline cmd/describe/machinepool/nodepool.go:76]: I'd review if the others "describe" commands have also conditional rendering.

Just to use the same pattern (show always the info with empty spaces vs only show it when is present).

Comment (andreadecorte) [reply cmd/describe/machinepool/nodepool.go:76]: yes, this is the same pattern we are using in the `describe cluster`, see https://github.com/openshift/rosa/blob/master/cmd/describe/cluster/cmd.go#L363-L369

File: pkg/ocm/machinepools.go [hunk 6]
```diff
@@ -32,6 +32,22 @@ func (c *Client) GetMachinePools(clusterID string) ([]*cmv1.MachinePool, error)
...
 }
 
+func (c *Client) GetMachinePool(clusterID string, machinePoolID string) (*cmv1.MachinePool, bool, error) {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).
+		MachinePools().
+		MachinePool(machinePoolID).
+		Get().
+		Send()
+	if response.Status() == 404 {
```
[the remark this answers is not in the document]
Comment (andreadecorte) [reply pkg/ocm/machinepools.go:42]: thanks fixed

File: cmd/describe/machinepool/machinepool.go (hunk 1 above)
Comment (andreadecorte) [reply cmd/describe/machinepool/machinepool.go]: I think we are not totally consistent, but yes let's converge on returning an error and exit code 1 for not found case.
example
```
rosa describe machinepool -c igal-rosa-test6 --machinepool workers
E: Machine pool 'workers' not found
```

File: cmd/describe/machinepool/nodepool.go (hunk 3 above)
Comment (andreadecorte) [reply cmd/describe/machinepool/nodepool.go]: changed

**Comment by oriAdler** [comment]
LGTM

**Comment by lucasponce** [comment]
> updated, example of updated output

Could you test the autoscaling case?

Just to confirm that the min/max are rendered in some way like that:

```
ID       AUTOSCALING  REPLICAS  INSTANCE TYPE  LABELS    TAINTS    AVAILABILITY ZONE  SUBNET                    VERSION  AUTOREPAIR
workers  Yes           [1-2]/2       m6a.xlarge                         us-east-1d         subnet-02e64da0728c6a969  4.12.21  Yes
```

**Comment by andreadecorte** [comment]
> > updated, example of updated output
> 
> Could you test the autoscaling case?
> 
> Just to confirm that the min/max are rendered in some way like that:
> 
> ```
> ID       AUTOSCALING  REPLICAS  INSTANCE TYPE  LABELS    TAINTS    AVAILABILITY ZONE  SUBNET                    VERSION  AUTOREPAIR
> workers  Yes           [1-2]/2       m6a.xlarge                         us-east-1d         subnet-02e64da0728c6a969  4.12.21  Yes
> ```

```
rosa list machinepool -c ad-int1
ID           AUTOSCALING  REPLICAS  INSTANCE TYPE  LABELS    TAINTS    AVAILABILITY ZONE  SUBNET                    VERSION  AUTOREPAIR
workers      No           2/2       m5.xlarge                          us-west-2a         subnet-0e3a4046c1c2f1078  4.12.31  Yes
autoscaling  Yes          0/1-2     m5.xlarge      aa=bb               us-west-2a         subnet-0e3a4046c1c2f1078  4.12.29  Yes
uhc-clusters-service/rosa % ./rosa describe machinepool -c ad-int1 autoscaling

ID:                         autoscaling
Cluster ID:                 267hmftbivfeak3q0lgrteg818g9luog
Autoscaling:                Yes
Desired replicas:           1-2
Current replicas:           0
Instance type:              m5.xlarge
Labels:                     aa=bb
Taints:
Availability zone:          us-west-2a
Subnet:                     subnet-0e3a4046c1c2f1078
Version:                    4.12.29
Autorepair:                 Yes
Tuning configs:
Message:                    WaitingForAvailableMachines: InstanceProvisionStarted,WaitingForNodeRef
```

**Pull request merged**